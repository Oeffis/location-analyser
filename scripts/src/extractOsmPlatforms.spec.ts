import { mkdir, readFile, writeFile } from "fs/promises";
import { beforeAll, expect, suite, test } from "vitest";
import { ExtractionResult, Node, OsmExtractor, Relation, Way } from "./osmExtractor";

suite("extractOsmPlatforms", () => {
    const BusStopRheinelbestraße = 6107133039;
    const TramStopRheinelbestraße = 125776045;
    const TrainStationGelsenkirchenPlatforms4and5 = 4250656;

    let extractor: OsmExtractor;
    let extraction: ExtractionResult;

    beforeAll(async () => {
        extractor = OsmExtractor.forPlatforms();

        try {
            const cachedExtraction = await readFile("./test-cache/extractedPlatforms.json", "utf-8");
            console.log("using cached extraction");

            const parsedCachedExtraction = JSON.parse(cachedExtraction) as { relations: Relation[]; ways: Way[]; nodes: Node[] };
            extraction = {
                relations: new Map(parsedCachedExtraction.relations.map((relation: Relation) => [relation.id, relation])),
                ways: new Map(parsedCachedExtraction.ways.map((way: Way) => [way.id, way])),
                nodes: new Map(parsedCachedExtraction.nodes.map((node: Node) => [node.id, node]))
            };
        } catch (e) {
            console.warn("no cached extraction found, extracting from OSM");
            console.error(e);
            extraction = await extractor.extract();
            await mkdir("./test-cache", { recursive: true });
            await writeFile("./test-cache/extractedPlatforms.json", JSON.stringify({
                relations: Array.from(extraction.relations.values()),
                ways: Array.from(extraction.ways.values()),
                nodes: Array.from(extraction.nodes.values())
            }));
        }
    }, 60000);

    test("extracts simple bus stop", () => {
        const node = extraction.nodes.get(BusStopRheinelbestraße);

        expect(node).not.toBeUndefined();
        expect(node).toMatchSnapshot();
    });

    test("extracts simple tram platform", () => {
        const way = extraction.ways.get(TramStopRheinelbestraße);

        expect(way).not.toBeUndefined();
        expect(way).toMatchSnapshot();
        for (const nodeId of way?.refs ?? []) {
            const node = extraction.nodes.get(nodeId);
            expect(node).not.toBeUndefined();
            expect(node).toMatchSnapshot();
        }
    });

    test("extracts outer section of a train station platform", () => {
        const relation = extraction.relations.get(TrainStationGelsenkirchenPlatforms4and5);

        expect(relation).not.toBeUndefined();
        expect(relation).toMatchSnapshot();

        for (const wayId of relation?.members.filter(member => member.role === "outer").map(m => m.ref) ?? []) {
            const way = extraction.ways.get(wayId);

            expect(way, `Way ${wayId} not found`).not.toBeUndefined();
            expect(way).toMatchSnapshot();
            for (const nodeId of way?.refs ?? []) {
                const node = extraction.nodes.get(nodeId);
                expect(node).not.toBeUndefined();
                expect(node).toMatchSnapshot();
            }
        }

        for (const wayId of relation?.members.filter(member => member.role === "inner").map(m => m.ref) ?? []) {
            expect(extraction.ways.has(wayId)).toBeFalsy();
        }
    });
});
