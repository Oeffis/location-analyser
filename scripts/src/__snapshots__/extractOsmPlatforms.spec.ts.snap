// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 1`] = `
{
  "id": 4250656,
  "members": [
    {
      "ref": 315907141,
      "role": "inner",
      "type": "way",
    },
    {
      "ref": 320974798,
      "role": "inner",
      "type": "way",
    },
    {
      "ref": 230235159,
      "role": "outer",
      "type": "way",
    },
  ],
  "tags": {
    "description": "Gelsenkirchen Hauptbahnhof, Bahnsteig 5+4",
    "height": "0.76",
    "layer": "2",
    "level": "2",
    "local_ref": "4;5",
    "name": "Gelsenkirchen Hbf",
    "public_transport": "platform",
    "railway": "platform",
    "ref": "4;5",
    "tactile_paving": "yes",
    "train": "yes",
    "type": "multipolygon",
    "wheelchair": "yes",
  },
  "type": "relation",
}
`;

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 2`] = `
{
  "id": 230235159,
  "refs": [
    454217531,
    1215879965,
    1215879908,
    2388241516,
    5038572467,
    454217532,
    454217533,
    454217534,
    454217536,
    454217537,
    5038572466,
    2388241500,
    454217538,
    454217539,
    454217530,
    454217531,
  ],
  "tags": {
    "level": "2",
  },
  "type": "way",
}
`;

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 3`] = `
{
  "id": 454217531,
  "lat": 51.5046342,
  "lon": 7.1006935,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 4`] = `
{
  "id": 1215879965,
  "lat": 51.504706,
  "lon": 7.1010485,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 5`] = `
{
  "id": 1215879908,
  "lat": 51.5047347,
  "lon": 7.101034,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 6`] = `
{
  "id": 2388241516,
  "lat": 51.5049401,
  "lon": 7.1020459,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 7`] = `
{
  "id": 5038572467,
  "lat": 51.505084,
  "lon": 7.1027538,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 8`] = `
{
  "id": 454217532,
  "lat": 51.5053097,
  "lon": 7.1038591,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 9`] = `
{
  "id": 454217533,
  "lat": 51.5058412,
  "lon": 7.1064241,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 10`] = `
{
  "id": 454217534,
  "lat": 51.5057852,
  "lon": 7.1064546,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 11`] = `
{
  "id": 454217536,
  "lat": 51.5053964,
  "lon": 7.1047479,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 12`] = `
{
  "id": 454217537,
  "lat": 51.5052248,
  "lon": 7.1039268,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 13`] = `
{
  "id": 5038572466,
  "lat": 51.5049937,
  "lon": 7.1028012,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 14`] = `
{
  "id": 2388241500,
  "lat": 51.5048476,
  "lon": 7.1020934,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 15`] = `
{
  "id": 454217538,
  "lat": 51.5046257,
  "lon": 7.1010172,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 16`] = `
{
  "id": 454217539,
  "lat": 51.5046478,
  "lon": 7.1010054,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 17`] = `
{
  "id": 454217530,
  "lat": 51.5045887,
  "lon": 7.1007162,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts outer section of a train station platform 18`] = `
{
  "id": 454217531,
  "lat": 51.5046342,
  "lon": 7.1006935,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts simple bus stop 1`] = `
{
  "id": 6107133039,
  "lat": 51.501452,
  "lon": 7.105743,
  "tags": {
    "bus": "yes",
    "highway": "bus_stop",
    "name": "Rheinelbestraße",
    "public_transport": "platform",
    "shelter": "yes",
  },
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts simple tram platform 1`] = `
{
  "id": 125776045,
  "refs": [
    1396187856,
    1396187866,
    2501560574,
    1396187880,
    4952536446,
    4952536445,
    4952536444,
    1396187856,
  ],
  "tags": {
    "name": "Rheinelbestraße",
    "public_transport": "platform",
    "railway": "platform",
    "tram": "yes",
  },
  "type": "way",
}
`;

exports[`extractOsmPlatforms > extracts > extracts simple tram platform 2`] = `
{
  "id": 1396187856,
  "lat": 51.5014985,
  "lon": 7.1060283,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts simple tram platform 3`] = `
{
  "id": 1396187866,
  "lat": 51.501604,
  "lon": 7.1058787,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts simple tram platform 4`] = `
{
  "id": 2501560574,
  "lat": 51.5017633,
  "lon": 7.1056881,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts simple tram platform 5`] = `
{
  "id": 1396187880,
  "lat": 51.5018069,
  "lon": 7.105636,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts simple tram platform 6`] = `
{
  "id": 4952536446,
  "lat": 51.501819,
  "lon": 7.1056639,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts simple tram platform 7`] = `
{
  "id": 4952536445,
  "lat": 51.5016136,
  "lon": 7.10592,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts simple tram platform 8`] = `
{
  "id": 4952536444,
  "lat": 51.5015176,
  "lon": 7.1060581,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > extracts > extracts simple tram platform 9`] = `
{
  "id": 1396187856,
  "lat": 51.5014985,
  "lon": 7.1060283,
  "type": "node",
}
`;

exports[`extractOsmPlatforms > transforms > transforms platforms 1`] = `
{
  "platformBounds": "id,lat,lon
6107133039,51.501452,7.105743
",
  "platforms": "id,name
6107133039,Rheinelbestraße
",
}
`;
